cloudinit/config/cc_grub_dpkg.py

```diff
@@ -62,22 +67,35 @@ def handle(name, cfg, _cloud, log, _args):
     idevs_empty = util.get_cfg_option_str(
         mycfg, "grub-pc/install_devices_empty", None)
 
-    if ((os.path.exists("/dev/sda1") and not os.path.exists("/dev/sda")) or
-       (os.path.exists("/dev/xvda1") and not os.path.exists("/dev/xvda"))):
+    # get the root disk where the /boot directory resides.
+    # the ESP will reside on the same root disk as /boot, but not device.
+    try:
+        disk = util.subp(['grub-probe', '-t', 'disk', '/boot'],
+                         capture=True)[0].strip()
+        # if /dev/disk/by-id/ exists, attempt to determine robust name
+        devlist = []
+        with suppress(FileNotFoundError):
+            for device in os.listdir("/dev/disk/by-id/"):
+                device = "{}{}".format("/dev/disk/by-id/", device)
+                devlist.append((device, str(Path(device).resolve())))
```

Comment:
Following on from the above, in pre-3.6 Pythons, [`.resolve()`](https://docs.python.org/3/library/pathlib.html#pathlib.Path.resolve) is strict and so will raise `FileNotFoundError` if the target path doesn't exist. Leaving the `suppress` in place would mean that we would ignore that error and proceed with an incomplete devlist.

Reply:
The latest revision has changed to udevadm and we no longer use `.resolve()` or `suppress` at all now, and their imports have been removed.